### PR TITLE
Don't retransmit a PipeDuplexRequestBody

### PR DESCRIPTION
**Pull request by swankjesse**

### CODE REVIEW

File: wire-grpc-tests/src/test/java/com/squareup/wire/GrpcClientTest.kt [hunk 1]
```diff
@@ -274,6 +275,41 @@ class GrpcClientTest {
...
+    mockService.enqueueSendFeature(name = "tree at 5,6")
+    mockService.enqueue(SendCompleted)
+    val grpcCall = routeGuideService.GetFeature()
+    val feature = grpcCall.executeBlocking(Point(latitude = 5, longitude = 6))
+    assertThat(feature).isEqualTo(Feature(name = "tree at 5,6"))
+
+    // Give the mock server invalid actions (SendCompleted with no response message), which will
+    // cause it to reset the socket connection.
+    mockService.enqueue(ReceiveCall("/routeguide.RouteGuide/RecordRoute"))
+    mockService.enqueue(SendCompleted)
```
Comment (carranca) [inline wire-grpc-tests/src/test/java/com/squareup/wire/GrpcClientTest.kt:300]: A-ha, nice and cleaner than the initial attempts.

File: wire-grpc-tests/src/test/java/com/squareup/wire/GrpcClientTest.kt [hunk 2]
```diff
@@ -274,6 +275,41 @@ class GrpcClientTest {
...
+
+    // Confirm we get an IOException from the broken connection and not an IllegalStateException
+    // from incorrectly attempting to recover.
+    val (requestChannel, deferredResponse) = routeGuideService.RecordRoute().executeBlocking()
+    val e = assertFailsWith<IOException> {
```
Comment (carranca) [inline wire-grpc-tests/src/test/java/com/squareup/wire/GrpcClientTest.kt:305]: Did you confirm that without your fix this was producing `IllegalStateException`?

**Review by carranca — APPROVED** [review]
Nice! Love the inline explanations too. Thanks for looking into this!

File: wire-grpc-client/src/jvmMain/kotlin/com/squareup/wire/internal/PipeDuplexRequestBody.kt [hunk 3]
```diff
@@ -39,4 +39,7 @@ internal class PipeDuplexRequestBody(
   }
 
   override fun isDuplex() = true
+
+  /** [Pipe.fold] can only be called once. */
+  override fun isOneShot() = true
```
Comment (oldergod) [inline wire-grpc-client/src/jvmMain/kotlin/com/squareup/wire/internal/PipeDuplexRequestBody.kt:44]: good find

**Review by oldergod — APPROVED** [review]

File: wire-grpc-tests/src/test/java/com/squareup/wire/GrpcClientTest.kt (hunk 2 above)
Comment (swankjesse) [reply wire-grpc-tests/src/test/java/com/squareup/wire/GrpcClientTest.kt:305]: Yes!

**Comment by rndhacker** [comment]
When is this change shipping to a new version? @oldergod

**Pull request merged**